marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetNextSteps.hs

```diff
@@ -0,0 +1,430 @@
+import Language.Marlowe.Runtime.ChainSync.Api
+  ( Address(Address)
+  , AssetId(AssetId)
+  , Assets(..)
+  , BlockHeader(..)
+  , BlockHeaderHash(..)
+  , ChainPoint
+  , PolicyId(PolicyId)
+  , TokenName(TokenName)
+  , Tokens(Tokens)
+  , TxId(..)
+  , TxOutRef(..)
+  , WithGenesis(..)
+  , fromDatum
+  )
+import Language.Marlowe.Runtime.Core.Api
+  ( ContractId(..)
+  , MarloweVersion(..)
+  , MarloweVersionTag(V1)
+  , Payout(..)
+  , Transaction(..)
+  , TransactionOutput(..)
+  , TransactionScriptOutput(..)
+  )
+import Language.Marlowe.Runtime.History.Api (ContractStep(..), RedeemStep(..))
+import Language.Marlowe.Runtime.Sync.Database (NextSteps(..))
+import qualified Plutus.V2.Ledger.Api as PV2
+import Prelude hiding (init)
+import Witherable (catMaybes, mapMaybe)
+
+getNextSteps :: MarloweVersion v -> ContractId -> ChainPoint -> T.Transaction (NextSteps v)
+getNextSteps MarloweV1 contractId point = do
+  orient point >>= \case
+    RolledBack toPoint -> pure $ Rollback toPoint
+    AtTip -> pure Wait
+    BeforeTip -> getNextTxIds contractId point >>= \case
+      Nothing -> pure Wait
+      Just NextTxIds{..} -> do
+        applySteps <- getApplySteps nextBlock contractId nextApplyTxIds
+        redeemSteps <- getRedeemSteps nextWithdrawalTxIds
+        pure $ Next nextBlock $ (ApplyTransaction <$> applySteps) <> (RedeemPayout <$> redeemSteps)
+
+data Orientation
+  = BeforeTip
+  | AtTip
+  | RolledBack ChainPoint
+
+orient :: ChainPoint -> T.Transaction Orientation
+orient Genesis = pure BeforeTip
+orient (At BlockHeader{..}) = T.statement (unBlockHeaderHash headerHash) $ decodeResult <$>
+  [maybeStatement|
+    SELECT
+      rollbackBlock.slotNo :: bigint?,
+      rollbackBlock.id :: bytea?,
+      rollbackBlock.blockNo :: bigint?
+    FROM marlowe.block
+    LEFT JOIN marlowe.block AS rollbackBlock ON block.rollbackToBlock = rollbackBlock.id
+    WHERE block.id = $1 :: bytea
+  |]
+  where
+    decodeResult Nothing = RolledBack Genesis
+    decodeResult (Just (Just slot, Just hash, Just block)) = RolledBack $ At BlockHeader
+      { slotNo = fromIntegral slot
+      , headerHash = BlockHeaderHash hash
+      , blockNo = fromIntegral block
+      }
+    decodeResult _ = BeforeTip
+
+data NextTxIds = NextTxIds
+  { nextBlock :: BlockHeader
+  , nextApplyTxIds :: [TxId]
+  , nextWithdrawalTxIds :: [TxId]
+  }
+
+getNextTxIds :: ContractId -> ChainPoint -> T.Transaction (Maybe NextTxIds)
+getNextTxIds (ContractId TxOutRef{..}) point = T.statement params $ fmap (NextTxIds <$> decodeBlockHeader <*> decodeApplyTxIds <*> decodeWithdrawalTxIds) <$>
+  [maybeStatement|
+    WITH params (createTxId, createTxIx, afterSlot) AS
+      ( SELECT $1 :: bytea, $2 :: smallint, $3 :: bigint
+      )
+    , nextApplyTxIds (slotNo, blockHeaderHash, blockNo, txIds) AS
+      ( SELECT
+          block.slotNo,
+          (ARRAY_AGG(block.id))[1],
+          (ARRAY_AGG(block.blockNo))[1],
+          ARRAY_AGG(applyTx.txId)
```

Comment:
I'm not sure if it's better to use `ARRAY_AGG` or just to add to the `GROUP BY`, given the `LIMIT 1`. I guess that either way is fine.

Reply:
I tried both - `(ARRAY_AGG(expr))[1]` gives better performance than multiple `GROUP BY` keys.